### DISCUSSION TITLE
Can I have a bundle with NPM packages?

### DISCUSSION THREAD

**Discussion by sandeepantony** [getporter/porter#1843] [Q&A] [answered]
I came across this tool while looking for the best practice to deploy an application built in microservice model. Our current deployment model uses Gitlab pipelines to deploy heterogenous artifacts to heterogenous resources. E.g we've npm packages deployed to AWS lambda, S3 and containers deployed to ECS. While the latter is solved by this tool, I'm curious to know if the former (npm packages to Lambda/S3) is a valid use case for this tool to solve.

**Comment by carolynvs** [comment]
Bundles are good for combining logic (scripts) and artifacts so that you can redistribute them for others to use. It kind of sounds like publishing npm packages to S3 is really more of a step in your build pipeline than an action that others would need to perform as well. So you may see limited utility in putting that into a bundle, but I may be misunderstanding.

**Reply by sandeepantony** [reply]
Let me elaborate. I've an application that is composed of 3 types of s/w to deploy... 
- docker image (container registry). Deployed to ECS
- npm package (npm registry), extracted and deployed to lambda
- npm package (npm registry), extracted and deployed to S3
Yes, all 3 will have its unique deploy step in the pipeline. But before the deploy step I wanted to create a release bundle that specifies the unique version of each s/w to deploy that I can distribute, e.g. to DevOps team that has access to artifact registry. So they can independently manage the pipelines to deploy to stage/prod. 

The concept is very close to cnab, except deployment artifacts are not just docker/oci images but also npm packages.

**Reply by carolynvs** [reply]
Oh! Thanks for clarifying a bit. Yes, this totally is something that bundles were intended to assist with. 

So when you installed a bundle for your app it would:

1. Deploy the docker image
2. Extract the package from npm and deploy to both lambda and s3

One thing that bundles can do is ship files as well. If you were interested in making a bundle that would still work if npm was down, you could include the npm package in the bundle's files as well and skip the npm step (or just not rely on it when deploying from a bundle). Same for the docker image, bundles can include images inside as well.

Sometimes we don't care about relying on things like a docker/npm registry at deploy time, so either way is a valid bundle technique (packaging everything in the bundle, or using the registry when the bundle is installed).

**Reply by sandeepantony** [reply]
Thanks! that helps!

**Comment by carolynvs** [comment, marked as answer]
Linking to the answer since GitHub discussions won't let me mark a nested comment as the answer

https://github.com/getporter/porter/discussions/1843#discussioncomment-1949389
